src/librustc_typeck/astconv.rs

```diff
@@ -110,7 +110,7 @@ pub trait AstConv<'tcx> {
     }
 
     /// What type should we use when a type is omitted?
-    fn ty_infer(&self, span: Span) -> Ty<'tcx>;
+    fn ty_infer(&self, default: Option<Ty<'tcx>>, span: Span) -> Ty<'tcx>;
```

Comment:
I think I would prefer to pass in an `Option<TypeParameterDef>`, or maybe some kind of more interesting "origin" thing. This would imply a default, but would give us richer information for error reporting and so forth.